### PR TITLE
bpo-43693: Add _PyCode_New() and do some related cleanup.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
The key change here is the addition of `_PyCode_New()`, which is an internal-only API to replace use of `PyCode_NewWithPosOnlyArgs()`.  The problem with that older API is the it has so many parameters.  This becomes especially painful when modifying its signature.  The new API uses a single struct to hold the values needed to create a new `PyCodeObject` instance.

Other changes:

* added `co_nfastlocals`, `co_ncellvars`, and `co_nfreevars`; this matters because I plan on replacing `co_varnames`, etc. with a single consolidated tuple
* moved some of the code tied to the internal resolution of `co_varnames`, etc. from ceval.c to codeobject.c (as functions); this reduced coupling makes it easier to change how `PyCodeObject` works internally
* re-arranged the `PyCodeObject` fields so they are grouped a bit more logically

There is still code in ceval.c (`_PyEval_MakeFrameVector()` specifically, and pyframeobject.c that is coupled to the internal structure of PyCodeObject, but this PR is big enough already. 🙂

https://bugs.python.org/issue43693

### CODE REVIEW

**Comment by markshannon** [comment]
I agree that we should add a non-API function for creating code objects and `_PyCodeNew` seems like a sensible name for it.
But surely that should be a 10 line diff, not a 1000 line diff?

Passing a struct as a single parameter to a function means that you have to ensure that all future versions can handling trailing zeroes.
E.g. 
```C
PyObject *make_thing(struct thing *t);
```
Version 1:
```
struct thing {
    some data;
};
```
Version 2:
```
struct thing {
   some data;
   int super_important_must_never_be_zero;
```
Any client code written for version 1 will compile for version 2 and then crash mysteriously when run.

**Comment by ericsnowcurrently** [comment]
I see your point but consider it less of an issue for an internal-only API like this.  For the most part this PR is about addressing the pain points I ran into while making changes around PyCodeObject:

* changes to the public "constructor" API to accommodate changes to internal details are frustrating for everyone (hence the new internal function)
* much of the public API for PyCodeObject probably shouldn't be public in the first place
* the signature of 2 of the "constructor" functions are particularly large, so adding, removing, or re-arranging parameters is annoying (hence the args struct, which we also do for frames)
* for folks not already familiar with PyCodeObject, the nature of the various fields isn't obvious at first (hence grouping the fields)
* more than a couple modules are fairly tightly coupled to the internal details of PyCodeObject, which may make sense for ceval.c as the confluence of functions, code, and frames, but that coupling (even in ceval,c) means changes to PyCodeObject internals result in touches to a variety of other modules (and the number of modules touched for a change correlates to bug count)

I'm okay with dialing back the changes if you think they aren't worth it, but the definitely are based in my experience thus far with PyCodeObject.

**Comment by markshannon** [comment]
I think we all agree that the API for making PyCodeObjects shouldn't have been public in the first place. But it is :disappointed: 
I'm in favour of removing it, but that would need a proper discussion of python-dev.

The accessor macros,  `_PyCode_GET_INSTRUCTION` etc, seem useful, but could you make them inline functions.

Ultimately we want something like https://github.com/markshannon/peps/blob/pep-mappable-pyc-file/pep-066x.rst#runtime-objects
It is going to take a while to get there, but let's try and minimize the changes along the way.

File: Include/cpython/code.h [hunk 1]
```diff
@@ -17,51 +19,75 @@ typedef struct _PyOpcache _PyOpcache;
...
-    int co_kwonlyargcount;      /* #keyword only arguments */
-    int co_nlocals;             /* #local variables */
-    int co_stacksize;           /* #entries needed for evaluation stack */
+
+    /* Note that not all fields are used in either hash or comparisons. */
+
+    /* metadata */
+    PyObject *co_filename;      /* unicode (where it was loaded from) */
+    /* co_name is used in for both the hash and comparisions.  This is
+     * done to preserve the name and line number for tracebacks and
```
Comment (gvanrossum) [inline Include/cpython/code.h]: How does using co_name preserve the line number? Is something missing from the comment here?

File: Include/internal/pycore_code.h [hunk 2]
```diff
@@ -24,9 +27,117 @@ struct _PyOpcache {
     char optimized;
 };
 
+
+// XXX Can we do this with an enum instead?
+typedef unsigned char _PyFastLocalKind;
+#define CO_FAST_POSONLY     0x01
+#define CO_FAST_POSORKW     0x02
+#define CO_FAST_VARARGS     0x04
+#define CO_FAST_KWONLY      0x09
```
Comment (gvanrossum) [inline Include/internal/pycore_code.h]: 9? Really? Why not a power of 2 like all the others?

File: Objects/codeobject.c [hunk 3]
```diff
@@ -151,169 +163,343 @@ validate_and_copy_tuple(PyObject *tup)
...
+    co->co_argcount = con->argcount;
+    co->co_posonlyargcount = con->posonlyargcount;
+    co->co_kwonlyargcount = con->kwonlyargcount;
+
+    co->co_stacksize = con->stacksize;
+
+    Py_INCREF(con->exceptiontable);
+    co->co_exceptiontable = con->exceptiontable;
+
+    co->co_nfastlocals = nfastlocals;
```
Comment (gvanrossum) [inline Objects/codeobject.c:442]: Does making the various count fields `Py_ssize_t` increase the size of a code object much? If yes, you would have to range-check these values before assigning them.

File: Include/internal/pycore_code.h [hunk 4]
```diff
@@ -24,9 +27,117 @@ struct _PyOpcache {
...
+#define CO_FAST_LOCALONLY   0x20
+#define CO_FAST_CELL        0x40
+#define CO_FAST_FREE        0x80
+#define CO_FAST_LOCAL (CO_FAST_POSONLY | CO_FAST_POSORKW | CO_FAST_VARARGS | \
+                       CO_FAST_KWONLY | CO_FAST_VARKWARGS |                  \
+                       CO_FAST_LOCALONLY)
+#define CO_FAST_ANY (CO_FAST_LOCAL | CO_FAST_CELL | CO_FAST_FREE)
+
+
+struct _PyCodeConstructor {
```
Comment (gvanrossum) [inline Include/internal/pycore_code.h:49]: This name feels kind of awkward -- a "constructor" in common parlance is a function or method of some sort, not a data structure. Maybe `_PyCodeConstructionArgs` or `_PyCodeNewArgs`?

File: Include/cpython/code.h [hunk 5]
```diff
@@ -2,6 +2,8 @@
 #  error "this header file must not be included directly"
 #endif
 
+// XXX Much of the code here should move to pycore_code.h.
```
Comment (gvanrossum) [inline Include/cpython/code.h:5]: What should be moved? Is this just about the contents of the `PyCodeObject` struct?

File: Python/marshal.c [hunk 6]
```diff
@@ -1333,6 +1334,7 @@ r_object(RFILE *p)
             kwonlyargcount = (int)r_long(p);
             if (PyErr_Occurred())
                 goto code_error;
+            // XXX We no longer use nlocals.
             nlocals = (int)r_long(p);
```
Comment (gvanrossum) [inline Python/marshal.c]: Then why not just write
```
r_long(p);  // nlocals is no longer used
```
(Also why not just remove it from the marshal format and change the magic number?)

**Review by gvanrossum — COMMENTED** [review]
This looks like a reasonable improvement (though indeed it also looks like some things will soon be deleted again).

File: Include/internal/pycore_code.h (hunk 2 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_code.h]: That's a typo.  Good catch!

File: Objects/codeobject.c (hunk 3 above)
Comment (ericsnowcurrently) [reply Objects/codeobject.c:442]: They should be int.  I'll fix that.

File: Include/internal/pycore_code.h (hunk 4 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_code.h:49]: I borrowed from the name we use for the same purpose with `_PyFrame_New_NoTrack()` -- I agree that the name is awkward but was trying to stick with precedent. 🙂  I don't mind calling it something different.

File: Include/cpython/code.h (hunk 5 above)
Comment (ericsnowcurrently) [reply Include/cpython/code.h:5]: IMO, we should move anything that folks shouldn't even think about if they aren't dealing with the internals.  I'd say nearly everything there falls in that category.  I'd especially make `PyCodeObject` opaque in the public API, like it is in the limited API.

File: Python/marshal.c (hunk 6 above)
Comment (ericsnowcurrently) [reply Python/marshal.c]: I was planning on doing it in the next PR where I'm changing the magic number for something else.  I don't mind doing it here though.

Comment (gvanrossum) [reply Python/marshal.c]: That's fine then.

File: Include/cpython/code.h (hunk 5 above)
Comment (gvanrossum) [reply Include/cpython/code.h:5]: Hm, anybody looking at code objects from C code is probably dealing with internals... What would be legitimate reasons to look at some fields here? I guess Pyjion and PyDev?

File: Include/internal/pycore_code.h (hunk 4 above)
Comment (gvanrossum) [reply Include/internal/pycore_code.h:49]: Hm, okay, the precedent is fine then.

File: Include/cpython/code.h (hunk 1 above)
Comment (ericsnowcurrently) [reply Include/cpython/code.h]: This comment already existed.  I only moved it and tweaked it slightly.  Perhaps I misunderstood the original comment?

(I did just notice a typo: "in for".)

Comment (gvanrossum) [reply Include/cpython/code.h]: But you changed the sense of the comment -- previously it had "The rest aren't used in either hash or comparisons, except for co_name, used in both." That made sense because all the fields above the comment are used in hash and comparisons, which includes co_firstlineno, and then co_name for some reason comes a little later. Maybe it's worth marking up the per-field comments with a mark indicating that it's used in comparisons, and explain the meaning of the mark in a comment preceding all fields?

Comment (ericsnowcurrently) [reply Include/cpython/code.h]: Ah, I'll clarify that then.

File: Include/cpython/code.h (hunk 5 above)
Comment (ericsnowcurrently) [reply Include/cpython/code.h:5]: Yeah, agreed.  I'm not going to tackle this part right now though.  It's going to take at least some discussion on python-dev.

**Comment by markshannon** [comment]
Please remove construction of a CodeObject using a structure. As I explained [earlier](https://github.com/python/cpython/pull/26258#issuecomment-844965567) it isn't safe.
Nor does it do anything for clarity, IMO, it just adds bulk.

File: Include/cpython/code.h [hunk 7]
```diff
@@ -17,51 +19,90 @@ typedef struct _PyOpcache _PyOpcache;
...
+     * - co_firstlineno
+     * - co_code
+     * - co_consts
+     * - co_names
+     * - co_varnames
+     * - co_freevars
+     * - co_cellvars
+     *
+     * This is done to preserve the name and line number for tracebacks
+     * and debuggers; otherwise, constant de-duplication would collapse
```
Comment (bl-ue) [inline Include/cpython/code.h:41]: "Deduplicate" is actually a word, so there's no need for a hyphen here.
```suggestion
     * and debuggers; otherwise, constant deduplication would collapse
```

File: Include/cpython/code.h [hunk 8]
```diff
@@ -17,51 +19,90 @@ typedef struct _PyOpcache _PyOpcache;
...
-    //  * n > 0 means there is cache in co_opcache[n-1].
+    //  * n > 0 means there is cache in co_opcache.cache[n-1].
     unsigned char *co_opcache_map;
-    _PyOpcache *co_opcache;
     int co_opcache_flag;  // used to determine when create a cache.
     unsigned char co_opcache_size;  // length of co_opcache.
+
+    /* redundant data */
+    int co_nfastlocals;         /* #entries needed for locals
+                                   (nlocals + ncellvars + nfreevars( */
```
Comment (bl-ue) [inline Include/cpython/code.h]: Typo?
```suggestion
    int co_nfastlocals;         /* #entries needed for locals
                                   (nlocals + ncellvars + nfreevars) */
```

File: Include/internal/pycore_code.h [hunk 9]
```diff
@@ -24,9 +27,117 @@ struct _PyOpcache {
...
+#define CO_FAST_POSORKW     0x02
+#define CO_FAST_VARARGS     0x04
+#define CO_FAST_KWONLY      0x08
+#define CO_FAST_VARKWARGS   0x10
+#define CO_FAST_LOCALONLY   0x20
+#define CO_FAST_CELL        0x40
+#define CO_FAST_FREE        0x80
+#define CO_FAST_LOCAL (CO_FAST_POSONLY | CO_FAST_POSORKW | CO_FAST_VARARGS | \
+                       CO_FAST_KWONLY | CO_FAST_VARKWARGS |                  \
+                       CO_FAST_LOCALONLY)
```
Comment (bl-ue) [inline Include/internal/pycore_code.h]: Sorry if this is naive, but can't you condense this into two lines?
```suggestion
#define CO_FAST_LOCAL (CO_FAST_POSONLY | CO_FAST_POSORKW | CO_FAST_VARARGS | \
                       CO_FAST_KWONLY | CO_FAST_VARKWARGS | CO_FAST_LOCALONLY)
```

File: Objects/frameobject.c [hunk 10]
```diff
@@ -649,6 +649,7 @@ frame_dealloc(PyFrameObject *f)
 static inline Py_ssize_t
 frame_nslots(PyFrameObject *frame)
 {
+    //return frame->f_code->co_nfastlocals;
```
Comment (bl-ue) [inline Objects/frameobject.c]: Leftover?
```suggestion
```

**Review by bl-ue — COMMENTED** [review]
Hi, I found a few things worth taking a look at :)

**Comment by gvanrossum** [comment]
> Please remove construction of a CodeObject using a structure. As I explained [earlier](https://github.com/python/cpython/pull/26258#issuecomment-844965567) it isn't safe.

That only applies for public APIs. This is a purely internal API. The compiler will catch the case you worry about in that explanation.

> Nor does it do anything for clarity, IMO, it just adds bulk.

It allows us something not entirely unlike keyword arguments in C, which I find a big gain to readability.

**Comment by markshannon** [comment]
No, the compiler won't catch it. It is legal to leave trailing fields off a struct assignment.
It only works if NULL or 0 is a sensible value for all the fields. For this struct, that is not the case. Even if it were, we would have laid a trap for future developers adding new fields.

**Comment by ericsnowcurrently** [comment]
@markshannon, calling `check_code()` will catch the problem you described.  That said, I don't see this as a problem for this internal API.  As to readability.  It's what Guido said.  To elaborate,: the order doesn't matter and you can leave "args" off if you want to use the default.

Regardless, there isn't a lot of value to a long discussion about this.  If you feel strongly about this then I'll change it to a function with many parameters.

File: Include/internal/pycore_code.h (hunk 9 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_code.h]: Your question isn't naive.  The reason I did what I did isn't obvious.  My preference is to group similar items in code, usually by line.  In this case, the first five items are all kinds of arguments, while the last one isn't.  I would have put the first group all on one line if that line weren't so long.

I'll clarify this, one way or another.  Thanks for pointing it out.

File: Include/cpython/code.h (hunk 7 above)
Comment (ericsnowcurrently) [reply Include/cpython/code.h:41]: Thanks for pointing that out.  In this case "deduplication" is an unusual and less familiar word.  With the hyphen is also the way it was spelled in the original comment that I adapted.  So I'd rather leave it that way.

File: Include/cpython/code.h (hunk 8 above)
Comment (ericsnowcurrently) [reply Include/cpython/code.h]: I can't tell you how many times my fingers make this specific mistake. 🙂

File: Objects/frameobject.c (hunk 10 above)
Comment (ericsnowcurrently) [reply Objects/frameobject.c]: @markshannon, before b11a951f16 this function really was equivalent to `frame->f_code->co_nfastlocals`, but you changed it to include the frame's stack.  Was this intentional?  Also, are `frame_traverse()` and `frame_tp_clear()` doing extra work then?

Comment (ericsnowcurrently) [reply Objects/frameobject.c]: Mark explained that now we track builtins/globals/locals too, hence the difference.

File: Include/internal/pycore_code.h [hunk 11]
```diff
@@ -24,9 +27,128 @@ struct _PyOpcache {
...
+
+    /* used by the eval loop */
+    PyObject *exceptiontable;
+};
+
+// Using an "arguments struct" like this is helpful for maintainability
+// in a case such as this with many parameters.  It does bear a risk:
+// if the struct changes and callers are not updated properly then the
+// compiler will not catch problems (like a missing argument).  This can
+// cause hard-to-debug problems.  The risk is miticated by the use of
```
Comment (bl-ue) [inline Include/internal/pycore_code.h]: Was reading and found a little typo.
```suggestion
// cause hard-to-debug problems.  The risk is mitigated by the use of
```

File: Include/cpython/code.h [hunk 12]
```diff
@@ -17,51 +19,89 @@ typedef struct _PyOpcache _PyOpcache;
...
+    /* the code */
     PyObject *co_code;          /* instruction opcodes */
+    int co_firstlineno;         /* first source line number */
+    PyObject *co_linetable;     /* string (encoding addr<->lineno mapping) See
+                                   Objects/lnotab_notes.txt for details. */
+
+    /* used by the code */
     PyObject *co_consts;        /* list (constants used) */
     PyObject *co_names;         /* list of strings (names used) */
+
```
Comment (markshannon) [inline Include/cpython/code.h:59]: `co_code`, `co_consts` and `co_names` are the hottest fields in the code object. Could make sure that they are near the start of the object. Moving the metadata section after this section should do.

File: Include/internal/pycore_code.h [hunk 13]
```diff
@@ -24,9 +27,147 @@ struct _PyOpcache {
     char optimized;
 };
 
+
+// We would use an enum if C let us specify the storage type.
```
Comment (markshannon) [inline Include/internal/pycore_code.h:31]: You can still use an enum to define the values though, even if you store them in a char.

File: Include/internal/pycore_code.h [hunk 14]
```diff
@@ -24,9 +27,147 @@ struct _PyOpcache {
...
+
+static inline PyObject *
+_PyCode_GetFreevar(PyCodeObject *co, int index)
+{
+    assert(index >= 0 && index < co->co_nfreevars);
+    return PyTuple_GetItem(co->co_freevars, index);
+}
+
+static inline bool
+_PyCode_CodeIsValid(PyCodeObject *co)
```
Comment (markshannon) [inline Include/internal/pycore_code.h:133]: This doesn't check for validity, at all. A full validity check would be large and slow. Maybe rename to reflect what it does?

File: Include/internal/pycore_code.h [hunk 15]
```diff
@@ -24,9 +27,147 @@ struct _PyOpcache {
...
+{
+    return PyBytes_Check(co->co_code) &&
+           PyBytes_GET_SIZE(co->co_code) <= INT_MAX &&
+           PyBytes_GET_SIZE(co->co_code) % sizeof(_Py_CODEUNIT) == 0 &&
+           _Py_IS_ALIGNED(PyBytes_AS_STRING(co->co_code),
+                          sizeof(_Py_CODEUNIT));
+}
+
+static inline _Py_CODEUNIT *
+_PyCode_GetInstructions(PyCodeObject *co)
```
Comment (markshannon) [inline Include/internal/pycore_code.h:143]: This is going to be misleading with quickening, so could you drop it for now.
Plus these accessor functions have an annoying tendency to end up in the public API, historically.

File: Objects/codeobject.c [hunk 16]
```diff
@@ -151,169 +168,387 @@ validate_and_copy_tuple(PyObject *tup)
...
 
-PyCodeObject *
-PyCode_NewWithPosOnlyArgs(int argcount, int posonlyargcount, int kwonlyargcount,
-                          int nlocals, int stacksize, int flags,
-                          PyObject *code, PyObject *consts, PyObject *names,
-                          PyObject *varnames, PyObject *freevars, PyObject *cellvars,
-                          PyObject *filename, PyObject *name, int firstlineno,
-                          PyObject *linetable, PyObject *exceptiontable)
+static int
+check_code(struct _PyCodeConstructor *con, enum check_mode mode)
```
Comment (markshannon) [inline Objects/codeobject.c:175]: Using boolean parameters (and `check_mode` is a boolean) that change behavior is generally regarded as bad API design.
I think it would be better if the legacy form just cleared the exception then called `PyErr_BadInternalCall()`

Comment (markshannon) [inline Objects/codeobject.c:175]: Does this add much overhead?
It isn't immediately clear whether you are adding new checks, or just giving better errors for pre-existing checks.

File: Objects/codeobject.c [hunk 17]
```diff
@@ -333,55 +568,224 @@ PyCode_New(int argcount, int kwonlyargcount,
...
     return result;
 }
 
 
+/******************
+ * fast locals
+ ******************/
+
+bool
+_PyCode_HasFastlocals(PyCodeObject *co, _PyFastLocalKind kind)
```
Comment (markshannon) [inline Objects/codeobject.c:623]: What is this for? What does the `kind` parameter do?

File: Objects/frameobject.c [hunk 18]
```diff
@@ -613,7 +613,7 @@ frame_dealloc(PyFrameObject *f)
 
     /* Kill all local variables */
     if (f->f_localsptr) {
-        for (int i = 0; i < co->co_nlocalsplus+FRAME_SPECIALS_SIZE; i++) {
+        for (int i = 0; i < co->co_nfastlocals+FRAME_SPECIALS_SIZE; i++) {
```
Comment (markshannon) [inline Objects/frameobject.c:616]: Either this is wrong, or `co_nfastlocals` is misnamed.
This should clear the fastlocals, cell and free variables (plus the specials), not just the "fast" locals.

File: Objects/frameobject.c [hunk 19]
```diff
@@ -649,6 +649,9 @@ frame_dealloc(PyFrameObject *f)
 static inline Py_ssize_t
 frame_nslots(PyFrameObject *frame)
 {
+    assert(frame->f_valuestack - frame->f_localsptr ==
+           // fastlocals + builtins + globals + locals
+           frame->f_code->co_nfastlocals + 3);
```
Comment (markshannon) [inline Objects/frameobject.c:654]: OK, I see it is just misnamed, and the comment is misleading. Cell and free variables are also included.

File: Objects/frameobject.c [hunk 20]
```diff
@@ -917,146 +920,28 @@ PyFrame_New(PyThreadState *tstate, PyCodeObject *code,
     return f;
 }
 
-/* Convert between "fast" version of locals and dictionary version.
```
Comment (markshannon) [inline Objects/frameobject.c:920]: Why all these changes? This seems unrelated to adding `_PyCode_New()`.

File: Objects/typeobject.c [hunk 21]
```diff
@@ -8830,71 +8831,50 @@ static int
 super_init_without_args(PyFrameObject *f, PyCodeObject *co,
                         PyTypeObject **type_p, PyObject **obj_p)
 {
-    if (co->co_argcount == 0) {
+    if (!_PyCode_HasFastlocals(co, CO_FAST_POSONLY | CO_FAST_POSORKW)) {
```
Comment (markshannon) [inline Objects/typeobject.c:8834]: Why?  This is performance critical code, and this looks a *lot* slower.

File: Objects/codeobject.c [hunk 22]
```diff
@@ -333,55 +568,224 @@ PyCode_New(int argcount, int kwonlyargcount,
...
+        _PyCode_FastInfoFromOffset(co, oparg, pname, pkind);
+    }
+    else {
+        // For now, opargs for cells and free vars are separate from locals.
+        _PyCode_FastInfoFromOffset(co, oparg + co->co_nlocals, pname, pkind);
+    }
+}
+
+int
+_PyCode_OffsetFromIndex(PyCodeObject *co, int index, _PyFastLocalKind kind)
```
Comment (markshannon) [inline Objects/codeobject.c:718]: I don't like the use of this function everywhere.
If inlined, it should be no slower than the code it is replacing, but what is the point?
`index + co->co_nlocals` is clearer to me than `_PyCode_OffsetFromIndex(co, index, CO_FAST_CELL)`.
`index` is far, far clearer than ` `_PyCode_OffsetFromIndex(co, index, CO_FAST_LOCAL)`

File: Python/ceval.c [hunk 23]
```diff
@@ -4876,61 +4873,61 @@ get_exception_handler(PyCodeObject *code, int index, int *level, int *handler, i
...
 
     /* Create a dictionary for keyword parameters (**kwags) */
     PyObject *kwdict;
-    Py_ssize_t i;
     if (co->co_flags & CO_VARKEYWORDS) {
         kwdict = PyDict_New();
         if (kwdict == NULL)
             goto fail;
-        i = total_args;
+        int offset = _PyCode_OffsetFromIndex(co, total_args, CO_FAST_LOCAL);
```
Comment (markshannon) [inline Python/ceval.c:4889]: This is performance critical code and this is going to be slower unless `_PyCode_OffsetFromIndex` is inlined.

**Review by markshannon — CHANGES_REQUESTED** [review]
This PR is too big for me to be confident that I'm not missing something important.
Why does it need to be so large?

There are a lot of changes to frameobject.c, typeobject.c and ceval.c that seem unrelated.
Several of which look like they will degrade performance.

**Comment by ericsnowcurrently** [comment]
If this PR is too big to review well enough then I'll split it up.

**Comment by ericsnowcurrently** [comment]
I'm closing this in favor of #26364 and #26375.  There are more changes than those cover, but the rest can be dealt with (or abandoned) later.
